source/math/FloatVectorList.ooc

```diff
@@ -381,4 +382,76 @@ FloatVectorList: class extends VectorList<Float> {
 			result add(this[i] clamp(floor, ceiling))
 		result
 	}
+	findMaxIndex: func -> Int {
+		index := 0
+		for (i in 1 .. this count)
+			if (this[index] < this[i])
+				index = i
+		index
+	}
+	crossCorrelation: func (other: This, vectorLength: Int) -> Float {
+		upperPart := 0.0f
+		lowerPart1 := 0.0f
+		lowerPart2 := 0.0f
+		thisMean := this mean
+		otherMean := other mean
+		thisPointer := (this pointer as Float*)
+		otherPointer := (other pointer as Float*)
+		for (i in 0 .. vectorLength) {
+			thisValue := thisPointer[i]
+			otherValue := otherPointer[i]
+			thisDiff := (thisValue - thisMean)
+			otherDiff := (otherValue - otherMean)
+			upperPart += thisDiff * otherDiff
+			lowerPart1 += thisDiff * thisDiff
+			lowerPart2 += otherDiff * otherDiff
+		}
+		upperPart / sqrt(lowerPart1 * lowerPart2)
+	}
+	shift: func (offset: Int, filledValue := 0.0f) -> This {
+			result := This new(this count)
+			thisPointer := (this pointer as Float*)
+			for (i in 0 .. this count) {
+				index := i + offset
+				value := (index >= 0 && index < this count) ? thisPointer[index] : filledValue
+				result add(value)
+			}
+			result
+	}
+	// offset < 0 means it shifts to left
+	findCrossCorrelationOffset: func (other: This, offsetStart, offsetEnd: Int) -> CorrelationData {
+		crossCorrelation := This new(this count)
+		for (offset in offsetStart .. offsetEnd)
+			crossCorrelation add(this shift(offset) crossCorrelation(other, other count))
+
+		index := crossCorrelation findMaxIndex()
+		maxValue := crossCorrelation[index]
+		index = (index < (offsetEnd - offsetStart) / 2) ? (index + offsetStart) : (index - offsetEnd)
+
+		CorrelationData new(index, maxValue, crossCorrelation)
+	}
+	// input: how many points there will be between two origin points.
+	// this is a Linear interpolation way to have higher precision
+	interpolate: func (numberOfPoints: Int) -> This {
+		result := this
+		if (numberOfPoints < 0)
+			Debug raise("numberOfPoints should be greather than 0 in FloatVectorList interpolate!")
+		else if (numberOfPoints > 0 && this count > 1) {
+			result = This new(this count + numberOfPoints * (this count - 1))
+			thisPointer := (this pointer as Float*)
+			for (index1 in 0 .. this count - 1) {
+				result add(thisPointer[index1])
+				for (index2 in 1 .. numberOfPoints + 1)
+					result add(thisPointer[index1] + (thisPointer[index1 + 1] - thisPointer[index1]) * index2 / (numberOfPoints + 1))
+			}
+			result add(thisPointer[this count - 1])
+		}
+		result
+	}
+}
+CorrelationData: cover {
+	index: Int
+	maxValue: Float
+	crossCorrelation: FloatVectorList
+	init: func@ (=index, =maxValue, =crossCorrelation)
 }
```

Comment:
What do you mean you can't return more than 2 variables? `RasterYuvPlanar _allocate` returns 3.